src/cargo/ops/fix.rs

```diff
@@ -582,7 +600,16 @@ impl Default for PrepareFor {
 impl FixArgs {
     fn get() -> FixArgs {
         let mut ret = FixArgs::default();
-        ret.rustc = env::args_os().nth(1).map(PathBuf::from);
+        if env::var(CLIPPY_FIX_ENV).is_ok() {
+            ret.rustc = Some(util::config::clippy_driver());
```

Comment:
I don't think this is the right place to set the clippy driver.  This ends up running clippy on every dependency.

I mentioned [here](https://github.com/rust-lang/cargo/issues/7006#issuecomment-505069317) that I think it would be worthwhile to see if the "run for primary crates" can be redone for `fix` and `clippy`.  Either we can try to change it here in this PR, or do it in a separate PR.  If it is done separately, then I think the decision to run rustc vs clippy-driver needs to be move to `rustfix_and_fix` so it only applies to primary crates.

Reply:
I think the problem is I still don't really understand what you guys meant by a primary unit only wrapper. This doesn't even look like its messing with a wrapper, its just setting the binary it treats as rustc, and I figured [this](https://github.com/yaahallo/cargo/blob/master/src/cargo/ops/fix.rs#L220) was already handling the "dont run on dependencies" requirement, though I didn't really check ... :S 

I would also prefer to implement the "run for primary crates" logic, so I think the next step is to just discuss what that means so I know what to do.

Reply:
At the end of `fix_maybe_exec_rustc`, there is a fall-through for all other dependencies.  `args.rustc` gets used down there, too (which in this case would be clippy-driver, which we don't want).

So, looking closer, my thought is that `BuildConfig::rustc_wrapper` could be changed to run for primary crates only. It looks like these things have evolved independently over time, and it's not clear if it must work the way it does now, but I don't see why it should. The only users appears to be clippy and fix, and both of them only want to run against the primary crates. `fix` appears to do a bunch of dancing around whether or not something is primary, but in the end, non-primary units only fall through to the end of `fix_maybe_exec_rustc` to call rustc, so I don't see that it is necessary in any way.

"Primary" is defined as the units created by `generate_targets`.  There is a method `Context::is_primary_package` to determine if a unit is primary or not.

The `Compilation` struct will probably need to be reworked a little to support deciding which process to use.  I would also move the `rustfix_diagnostic_server` code in `Compilation::new` to the `fix` function in `fix.rs`, since it already has access to a `ProcessBuilder` (I think when this was originally written, it did not work that way).

Then I think the decision on which wrapper to use would go in [`prepare_rustc`](https://github.com/rust-lang/cargo/blob/143aaed0eedda82f6684637ba1bf2d69b24c7799/src/cargo/core/compiler/mod.rs#L591-L601).  Perhaps it could pass a bool to `Compilation::rustc_process` to tell it is a primary unit? Whatever seems cleanest here, I'm not sure.

[These lines](https://github.com/rust-lang/cargo/blob/143aaed0eedda82f6684637ba1bf2d69b24c7799/src/cargo/core/compiler/mod.rs#L183-L185) could then be removed. And all of the checks for `primary_package` in `fix.rs` could be removed.

Reply:
Doesn't rustc_wrapper get used by people who set it via the env variable, most notably sccache?

Reply:
`BuildConfig::rustc_wrapper` is only used by fix and clippy AFAIK.  RUSTC_WRAPPER gets set directly into [`Rustc`](https://github.com/rust-lang/cargo/blob/0bb6e14ee84ceeff8e74642420d8aae7e0deb9b1/src/cargo/util/rustc.rs#L23) [here](https://github.com/rust-lang/cargo/blob/0bb6e14ee84ceeff8e74642420d8aae7e0deb9b1/src/cargo/util/config.rs#L213-L216).

Reply:
wow yea, `rustc_wrapper` and `Rustc.wrapper` are two different variables... thats wild

Reply:
Ok I think I did it correctly.

Reply:
The main worry I have is that I misunderstood the superprocess subprocess relationship between the cargo invocations.

Theres all that stuff about primary and non-primary pass throughs in `fix_maybe_exec_rustc` but the solution I came to involved still using cargo as a wrapper and changing the rustc path that its wrapping if its a primary crate. The entire cargo subprocess of one of these primary crates will use clippy-driver as its rustc path, so if its doing things like building ITS dependencies which aren't primary even tho the super crate itself is a primary package this will still be broken.

Reply:
@yaahallo oh so currently there's only one Cargo and it spawns a bunch of rustc instances. In the case of cargo-fix some of those "rustc instances" actually happen to be the Cargo executable masquerading as rustc, and then that Cargo calls the actual rustc.

The primary/non-primary business is just that most compilations just need to go direct to rustc, but the "best" way for modeling that is to just always call cargo-as-a-wrapper which simply sometimes quickly delegates to rustc as if it were only rustc.

Or maybe I'm misundertanding the question here? I could be missing context...

Reply:
I think your answering the question, the main thing I'm unclear about is the scope of work cargo does when masquerading as rustc, like if it just does a little work invokes the compiler as specified by the primary cargo process and then captures the output and does a little more work then that makes sense and is what I would expect.

Reply:
For the case of cargo fix specifically the primary units need a significant amount of extra work done and all non-primary units don't need any amount of extra work done (theoretically at least).

The significant amount of work with respect to cargo fix is executing rustc multiple times to learn about fixes, apply them, maybe learn about more fixes, apply again, etc.

Reply:
yea but its just repeating the same job as necessary right? Not doing any more dependency resolution or calling rustc on more than one crate from within one of those cargo subprocess.